### PR TITLE
AT-2672 Visualizer apply

### PR DESCRIPTION
**Pull request by dariuscox**

### CODE REVIEW

File: terrawrap/models/graph.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+        self.command = command
+        self.reverse_pipeline = command == 'destroy'
+        self.graph = graph
+        self.graph_dict = {}
+        self.post_graph = post_graph
+        self.prefix = prefix
+        self.not_applied = set()
+        self.failures = []
+
+    def execute_graph(self, num_parallel: int = 4, debug: bool = False, print_only_changes: bool = False):
```
Comment (aoskotsky-amplify) [inline terrawrap/models/graph.py]: Codacy Issue found: [Too many local variables (17/15)](https://app.codacy.com/manual/amplify-education/terrawrap/pullRequest?prid=4904155)

**Comment by aoskotsky-amplify** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 10
           

Complexity increasing per file
==============================
- terrawrap/models/graph.py  15
- terrawrap/utils/graph.py  5
- terrawrap/utils/config.py  5
- terrawrap/models/graph_entry.py  9
- test/unit/test_graph.py  1
         

Clones added
============
- terrawrap/models/graph.py  5
- terrawrap/models/graph_entry.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/amplify-education/terrawrap/pullRequest?prid=4904155&bid=16239081)

File: terrawrap/models/graph.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+            else:
+                if item.state == "no-op":
+                    self.not_applied.add(item.path)
+        if self.not_applied:
+            print("The following directories have not been run since they are out of scope")
+            print(self.not_applied)
+
+        if self.failures:
+            raise RuntimeError(
+                "The follow directories failed with command '%s':\n%s" % (self.command, "\n".join(self.failures))
```
Comment (aoskotsky-amplify) [inline terrawrap/models/graph.py]: Codacy Issue found: [Line too long (113/110)](https://app.codacy.com/manual/amplify-education/terrawrap/pullRequest?prid=4904155)

File: terrawrap/models/graph_entry.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
+"""Module for containing Graph Entries"""
+import logging
+import os
+import tempfile
+import time
```
Comment (aoskotsky-amplify) [inline terrawrap/models/graph_entry.py]: Codacy Issue found: [Unused import time](https://app.codacy.com/manual/amplify-education/terrawrap/pullRequest?prid=4904155)

File: terrawrap/utils/graph.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+    for node in graph:
+        path = Path(node)
+        if path.is_symlink():
+            symlinks.append(path)
+
+    return symlinks
+
+
+def connect_symlinks(graph, symlink_dict):
```
Comment (aoskotsky-amplify) [inline terrawrap/utils/graph.py]: Codacy Issue found: [Missing function or method docstring](https://app.codacy.com/manual/amplify-education/terrawrap/pullRequest?prid=4904155)

File: terrawrap/models/graph.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
+"""Module containing the ApplyGraph class"""
+import concurrent.futures
+from typing import Iterable, Optional, List, DefaultDict
```
Comment (aoskotsky-amplify) [inline terrawrap/models/graph.py]: Codacy Issue found: [Unused DefaultDict imported from typing](https://app.codacy.com/manual/amplify-education/terrawrap/pullRequest?prid=4904155)

File: terrawrap/models/graph.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+        :param successors: A list of successors to be executed from the previous call
+        :param num_parallel: The number of pipeline entries to run in parallel.
+        :param debug: True if Terraform debugging should be turned on.
+        :param print_only_changes: True if only directories which contained changes should be printed.
+        """
+        futures_to_paths = {}
+
+        for node in successors:
+            entry = self._get_or_create_entry(node)
+            if entry.state is not "Pending":
```
Comment (aoskotsky-amplify) [inline terrawrap/models/graph.py]: Codacy Issue found: [Comparison to literal](https://app.codacy.com/manual/amplify-education/terrawrap/pullRequest?prid=4904155)

File: terrawrap/utils/graph.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
+""" Holds graph utilities"""
+
+import os
+from typing import List, Tuple, Any
+import networkx
+from pathlib import Path
```
Comment (aoskotsky-amplify) [inline terrawrap/utils/graph.py]: Codacy Issue found: [standard import "from pathlib import Path" should be placed before "import networkx"](https://app.codacy.com/manual/amplify-education/terrawrap/pullRequest?prid=4904155)

File: terrawrap/models/graph_entry.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+                capture_stderr=True,
+                env=command_env
+            )
+            operation_args += [plan_file_name]
+        else:
+            plan_exit_code = 0
+            plan_stdout = []
+
+        changes_detected = plan_exit_code != 0
+        if plan_exit_code == 0 or plan_exit_code == 2:
```
Comment (aoskotsky-amplify) [inline terrawrap/models/graph_entry.py]: Codacy Issue found: [Consider merging these comparisons with "in" to 'plan_exit_code in (0, 2)'](https://app.codacy.com/manual/amplify-education/terrawrap/pullRequest?prid=4904155)

File: terrawrap/models/graph_entry.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
+"""Module for containing Graph Entries"""
+import logging
+import os
+import tempfile
+import time
+import random
```
Comment (aoskotsky-amplify) [inline terrawrap/models/graph_entry.py]: Codacy Issue found: [Unused import random](https://app.codacy.com/manual/amplify-education/terrawrap/pullRequest?prid=4904155)

File: terrawrap/utils/graph.py [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+        for node in path:
+            tab_spacing = int(node[0]) - 1 + depth
+            if depth > 0:
+                print()
+            relative_node = node[1].replace(os.getcwd(), "")
+            print(("\t" * tab_spacing) + ">", relative_node)
+        depth += 1
+
+
+def find_symlink_directories(graph):
```
Comment (aoskotsky-amplify) [inline terrawrap/utils/graph.py]: Codacy Issue found: [Missing function or method docstring](https://app.codacy.com/manual/amplify-education/terrawrap/pullRequest?prid=4904155)

**Comment by aoskotsky-amplify** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- terrawrap/models/graph_entry.py  8
- test/unit/test_graph.py  1
- terrawrap/utils/graph.py  5
- terrawrap/utils/config.py  5
- terrawrap/models/graph.py  15
         

Clones added
============
- terrawrap/models/graph_entry.py  2
- terrawrap/models/graph.py  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/amplify-education/terrawrap/pullRequest?prid=4904155&bid=16239081)

**Comment by aoskotsky-amplify** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Coverage increased per file
===========================
+ terrawrap/utils/path.py  5
+ terrawrap/utils/collection_utils.py  19
+ terrawrap/utils/config.py  14
         

Complexity increasing per file
==============================
- terrawrap/utils/graph.py  5
- test/unit/test_graph_entry.py  1
- terrawrap/models/graph.py  15
- test/unit/test_apply_graph.py  1
- terrawrap/utils/config.py  5
- terrawrap/models/graph_entry.py  8
- test/unit/test_graph.py  1
         

Clones added
============
- test/unit/test_graph_entry.py  1
- terrawrap/models/graph.py  5
- terrawrap/models/pipeline_entry.py  2
- terrawrap/models/graph_entry.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/amplify-education/terrawrap/pullRequest?prid=4904155&bid=16239081)

File: test/unit/test_apply_graph.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+        graph_entry_class.return_value.state = "no-op"
+        graph_entry_class.return_value.path = "foo/app1"
+        graph_entry_class.return_value.execute.return_value = (0, ['Success'], True)
+
+        graph = ApplyGraph('plan', self.graph, self.post_graph, "bar")
+
+        graph.execute_graph()
+        graph.execute_post_graph()
+
+        expected_not_applied = {'foo/app1'}
```
Comment (aoskotsky-amplify) [inline test/unit/test_apply_graph.py]: Codacy Issue found: [Unused variable 'expected_not_applied'](https://app.codacy.com/manual/amplify-education/terrawrap/pullRequest?prid=4904155)

**Comment by aoskotsky-amplify** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Coverage increased per file
===========================
+ terrawrap/utils/path.py  5
+ terrawrap/utils/config.py  14
+ terrawrap/utils/collection_utils.py  19
         

Complexity increasing per file
==============================
- terrawrap/models/graph_entry.py  8
- terrawrap/models/graph.py  15
- test/unit/test_graph_entry.py  1
- terrawrap/utils/config.py  5
- terrawrap/utils/graph.py  5
- test/unit/test_apply_graph.py  1
- test/unit/test_graph.py  1
         

Clones added
============
- terrawrap/models/graph_entry.py  2
- terrawrap/models/graph.py  5
- terrawrap/models/pipeline_entry.py  2
- test/unit/test_graph_entry.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/amplify-education/terrawrap/pullRequest?prid=4904155&bid=16239081)

**Comment by aoskotsky-amplify** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Coverage increased per file
===========================
+ terrawrap/utils/path.py  5
+ terrawrap/utils/config.py  14
+ terrawrap/utils/collection_utils.py  19
         

Complexity increasing per file
==============================
- terrawrap/models/graph.py  15
- test/unit/test_graph_entry.py  1
- terrawrap/models/graph_entry.py  8
- test/unit/test_graph.py  1
- test/unit/test_apply_graph.py  1
- terrawrap/utils/config.py  5
- terrawrap/utils/graph.py  5
         

Clones added
============
- terrawrap/models/graph.py  5
- test/unit/test_graph_entry.py  1
- terrawrap/models/graph_entry.py  2
- terrawrap/models/pipeline_entry.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/amplify-education/terrawrap/pullRequest?prid=4904155&bid=16239081)

**Comment by aoskotsky-amplify** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Coverage increased per file
===========================
+ terrawrap/utils/config.py  14
+ terrawrap/utils/path.py  5
+ terrawrap/utils/collection_utils.py  19
         

Complexity increasing per file
==============================
- terrawrap/utils/config.py  5
- terrawrap/utils/graph.py  5
- test/unit/test_apply_graph.py  1
- test/unit/test_graph_entry.py  1
- test/unit/test_graph.py  1
- terrawrap/models/graph_entry.py  8
- terrawrap/models/graph.py  15
         

Clones added
============
- test/unit/test_graph_entry.py  1
- terrawrap/models/graph_entry.py  2
- terrawrap/models/graph.py  5
- terrawrap/models/pipeline_entry.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/amplify-education/terrawrap/pullRequest?prid=4904155&bid=16239081)

**Comment by aoskotsky-amplify** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- terrawrap/models/graph_entry.py  8
- test/unit/test_graph.py  1
- test/unit/test_graph_entry.py  1
- test/unit/test_apply_graph.py  1
- terrawrap/models/graph.py  15
- terrawrap/utils/config.py  5
- terrawrap/utils/graph.py  5
         

Clones added
============
- terrawrap/models/graph_entry.py  2
- test/unit/test_graph_entry.py  1
- terrawrap/models/graph.py  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/amplify-education/terrawrap/pullRequest?prid=4904155&bid=16239081)

**Comment by aoskotsky-amplify** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- test/unit/test_graph_entry.py  1
- test/unit/test_apply_graph.py  1
- test/unit/test_graph.py  1
- terrawrap/models/graph.py  15
- terrawrap/utils/graph.py  5
- terrawrap/utils/config.py  5
- terrawrap/models/graph_entry.py  8
         

Clones added
============
- test/unit/test_graph_entry.py  1
- terrawrap/models/graph.py  5
- terrawrap/models/graph_entry.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/amplify-education/terrawrap/pullRequest?prid=4904155&bid=16239081)

**Comment by aoskotsky-amplify** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Coverage increased per file
===========================
+ terrawrap/utils/path.py  5
+ terrawrap/utils/collection_utils.py  19
+ terrawrap/utils/config.py  14
         

Complexity increasing per file
==============================
- test/unit/test_apply_graph.py  1
- test/unit/test_graph.py  1
- terrawrap/models/graph.py  15
- test/unit/test_graph_entry.py  1
- terrawrap/utils/config.py  5
- terrawrap/models/graph_entry.py  8
- terrawrap/utils/graph.py  5
         

Clones added
============
- terrawrap/models/pipeline_entry.py  2
- terrawrap/models/graph.py  5
- test/unit/test_graph_entry.py  1
- terrawrap/models/graph_entry.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/amplify-education/terrawrap/pullRequest?prid=4904155&bid=16239081)

**Comment by aoskotsky-amplify** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Coverage increased per file
===========================
+ terrawrap/utils/config.py  14
+ terrawrap/utils/collection_utils.py  19
+ terrawrap/utils/path.py  5
         

Complexity increasing per file
==============================
- terrawrap/models/graph.py  15
- terrawrap/utils/config.py  5
- test/unit/test_graph.py  1
- terrawrap/utils/graph.py  5
- terrawrap/models/graph_entry.py  8
- test/unit/test_graph_entry.py  1
- test/unit/test_apply_graph.py  1
         

Clones added
============
- terrawrap/models/graph.py  5
- terrawrap/models/pipeline_entry.py  3
- terrawrap/models/graph_entry.py  3
- test/unit/test_graph_entry.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/amplify-education/terrawrap/pullRequest?prid=4904155&bid=16239081)

**Comment by aoskotsky-amplify** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Coverage increased per file
===========================
+ terrawrap/utils/collection_utils.py  19
+ terrawrap/utils/config.py  14
+ terrawrap/utils/path.py  5
         

Complexity increasing per file
==============================
- terrawrap/models/graph.py  15
- terrawrap/utils/config.py  5
- terrawrap/models/graph_entry.py  8
- test/unit/test_apply_graph.py  1
- test/unit/test_graph_entry.py  1
- test/unit/test_graph.py  1
- terrawrap/utils/graph.py  5
         

Clones added
============
- terrawrap/models/pipeline_entry.py  3
- terrawrap/models/graph.py  5
- terrawrap/models/graph_entry.py  3
- test/unit/test_graph_entry.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/amplify-education/terrawrap/pullRequest?prid=4904155&bid=16239081)

**Comment by aoskotsky-amplify** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Coverage increased per file
===========================
+ terrawrap/utils/config.py  14
+ terrawrap/utils/path.py  5
+ terrawrap/utils/collection_utils.py  19
         

Complexity increasing per file
==============================
- test/unit/test_graph.py  1
- test/unit/test_apply_graph.py  1
- test/unit/test_graph_entry.py  1
- terrawrap/utils/graph.py  5
- terrawrap/utils/config.py  5
- terrawrap/models/graph.py  15
- terrawrap/models/graph_entry.py  8
         

Clones added
============
- test/unit/test_graph_entry.py  1
- terrawrap/models/graph.py  5
- terrawrap/models/graph_entry.py  3
- terrawrap/models/pipeline_entry.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/amplify-education/terrawrap/pullRequest?prid=4904155&bid=16239081)

**Comment by aoskotsky-amplify** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Coverage decreased per file
===========================
- terrawrap/utils/graph.py  -20
         

Complexity increasing per file
==============================
- test/unit/test_apply_graph.py  1
- terrawrap/models/graph.py  15
- terrawrap/models/graph_entry.py  8
- test/unit/test_graph_entry.py  1
         

Clones added
============
- terrawrap/models/graph.py  5
- terrawrap/models/graph_entry.py  3
- test/unit/test_graph_entry.py  1
- terrawrap/models/pipeline_entry.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/amplify-education/terrawrap/pullRequest?prid=4904155&bid=16239081)

File: bin/graph_apply [hunk 12]
```diff
@@ -0,0 +1,110 @@
+#!/usr/bin/env python
+"""
+This script obtains the dependencies for the terraform config wrapper files and applies them.
+
+Usage:
+    graph_apply.py --path=PATH [--operation=OPERATION] [--parallel-jobs=NUM_JOBS] [--debug] [--print-only-changes]
```
Comment (dariuscox) [inline bin/graph_apply]: remove .py

File: bin/graph_apply [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+    graph_apply.py --version
+
+Arguments:
+    path    The path of the config directory to check
+
+Options:
+    -h, --help          Show this message and exit.[default: False]
+    -p PATH --path=PATH                         Path to the directory to execute.
```
Comment (dariuscox) [inline bin/graph_apply]: align whitespace

Comment (dariuscox) [reply bin/graph_apply]: add commas

File: bin/graph_apply [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+        config_dir = os.path.dirname(config_dir)
+
+    wrapper_config_dict = {}
+    print("Visualizing Dependencies for %s:" % config_dir)
+    try:
+        graph, post_graph = walk_and_graph_directory(config_dir, wrapper_config_dict)
+    except ValueError:
+        graph, post_graph = None, None
+        print("Graph apply can only be used for applying multiple directories, "
+              "please use tf apply for and individual directory")
```
Comment (dariuscox) [inline bin/graph_apply]: tf command not just apply

File: terrawrap/models/graph.py [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+    """Class for representing an Apply Graph."""
+
+    def __init__(self, command: str, graph: networkx.DiGraph, post_graph: List[str], prefix: str):
+        """
+        :param command: The Terraform command that this pipeline should execute.
+        :param pipeline_path: Path to the Terrawrap pipeline file. Can be absolute or relative from where
+        the script is executed.
+        """
+        self.command = command
+        self.reverse_pipeline = command == 'destroy'
```
Comment (dariuscox) [inline terrawrap/models/graph.py]: reverse graph

File: terrawrap/models/graph.py [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+                    exit_code, stdout, changes_detected = future.result()
+
+                    if print_only_changes and not changes_detected:
+                        stdout = ["No changes detected.\n"]
+
+                    print("\nFinished executing %s %s ..." % (path, self.command))
+                    print("Output:\n\n%s\n" % "".join(stdout).strip())
+
+                    if exit_code != 0:
+                        self.failures.append(path)
```
Comment (dariuscox) [inline terrawrap/models/graph.py]: print failures

File: terrawrap/utils/graph.py [hunk 17]
```diff
@@ -106,3 +107,38 @@ def visualize(dependencies: List[List[str]]):
...
+    :param graph: The graph containing symlinked directories
+    :param symlink_dict: A dictionary of symlinks and the directories that are linked to them.
+    """
+    for symlink in symlink_dict.keys():
+        current = symlink
+        secondary = 0
+        links = list(symlink_dict[symlink])
+        while secondary is not len(links):
+            secondary_link = links[secondary]
+            print("adding edge from %s to %s" % (current, secondary_link))
```
Comment (dariuscox) [inline terrawrap/utils/graph.py]: remove print

**Comment by aoskotsky-amplify** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Coverage decreased per file
===========================
- terrawrap/utils/graph.py  -19
         

Complexity increasing per file
==============================
- terrawrap/models/graph_entry.py  8
- test/unit/test_apply_graph.py  1
- terrawrap/models/graph.py  15
- test/unit/test_graph_entry.py  1
         

Clones added
============
- terrawrap/models/graph_entry.py  3
- terrawrap/models/pipeline_entry.py  3
- terrawrap/models/graph.py  5
- test/unit/test_graph_entry.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/amplify-education/terrawrap/pullRequest?prid=4904155&bid=16239081)

File: terrawrap/models/graph.py [hunk 18]
```diff
@@ -0,0 +1,236 @@
...
+            if not item:
+                self.not_applied.add(node)
+            else:
+                if item.state == "no-op":
+                    self.not_applied.add(item.path)
+        if self.not_applied:
+            print("The following directories have not been run since they are out of scope")
+            print(self.not_applied)
+
+        if self.failures:
```
Comment (jconstance-amplify) [inline terrawrap/models/graph.py]: Can we add some printout of all the directories that were skipped because their dependencies failed?

Comment (dariuscox) [reply terrawrap/models/graph.py]: Thats in the print statement above, if its skipped because of the entries above it will be added to not applied

Comment (jconstance-amplify) [reply terrawrap/models/graph.py]: Ah, okay. I wasn't sure if that was where that would end up.

**Pull request merged**